strawberryfields/program.py

```diff
@@ -702,21 +488,20 @@ def compile(self, backend='fock', **kwargs):
             raise ValueError("Could not find backend {} in Strawberry Fields database".format(backend))
 
         if db.modes is not None:
-            if self.num_subsystems > db.modes:
-                raise CircuitError("This program requires {} modes, but the {} backend "
-                                   "only supports a {} mode program".format(self.num_subsystems, backend, db.modes))
+            # subsystems may be created and destroyed, this is total number that has ever existed
```

Comment:
Compilation as such should be, and the topology-checking currently requires complete isomorphism so if the program has a `_New_modes` operation in it, the template has to have one too.

You can think of `_New_modes` and `Delete` diagrammatically as circuit elements/nodes which start and end wires in the middle of a circuit, respectively.

Reply:
Maybe _New_modes should be renamed to `_Create`, or _Delete to `_Delete_modes` for consistency.

Reply:
I think they used to be called `New` and `Del` before, didn't they? :laughing:

Reply:
Regarding above discussion, I suppose our topology checks would fail for two circuits which contain all the same gates, but differed only in the use of `New` and `Del`. Not a big deal, but something we can think about if we want a more robust topology-checking mechanism in future